### PR TITLE
JSON Editor For Currency options

### LINKED ISSUE #1844: Ability to edit Currency options from admin panel

### PR DESCRIPTION
**Pull request by dinesh-skyach**
JSON Editor For Currencies 
closes #1844

### CODE REVIEW

File: app/assets/javascripts/admin/json_editor.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        json_editor = new JSONEditor(document.getElementById('json_editor'), {
+            form_name_root: 'currency[options]',
+            disable_collapse: true,
+            theme: 'bootstrap3',
+            disable_properties: true,
+            iconlib: 'fontawesome4',
+            schema: {
+                title: 'Options',
+                type: 'object',
+                required: jQuery.parseJSON(gon.json_editor_required),
```
Comment (ysv) [inline app/assets/javascripts/admin/json_editor.js]: Don't use gon

File: app/views/admin/currencies/show.html.erb [hunk 2]
```diff
@@ -78,10 +67,28 @@
...
   <div class="row">
     <div class="col-12">
       <%= f.submit 'Submit', class: 'btn btn-primary' %>
     </div>
   </div>
 <% end %>
+
+<script>
+    $(document).ready(function() {
+      json_editor.setValue(jQuery.parseJSON('<%= raw Hash[ *Currency::OPTIONS_ATTRIBUTES.collect { |v| [ v, @currency.options[v] ] }.flatten ].to_json %>'));
```
Comment (ysv) [inline app/views/admin/currencies/show.html.erb]: We need ability to add custom attributes into options dynamicly

File: spec/controllers/admin/currencies_controller_spec.rb [hunk 3]
```diff
@@ -46,7 +47,8 @@
         enabled:                          false,
         base_factor:                      100000,
         precision:                        9,
-        erc20_contract_address:           '12kAmv8QXvQyosGzitFYm6YzxK2SgovhQ9' }
+        options:                          { "erc20_contract_address" => '12kamv8qxvqyosgzitfym6yzxk2sgovhq9'}
```
Comment (ysv) [inline spec/controllers/admin/currencies_controller_spec.rb]: add some custom option
for example 
```json
{ "erc20_contract_address":  "12kamv8qxvqyosgzitfym6yzxk2sgovhq9", "custom_currency_option": "options value"}
```

**Review by ysv — CHANGES_REQUESTED** [review]

**Comment by dinesh-skyach** [comment]
@ysv I have made some changes. Please have a look again.

File: app/controllers/admin/base_controller.rb [hunk 4]
```diff
@@ -8,6 +8,8 @@ class BaseController < ::ApplicationController
     before_action :auth_admin!
     before_action :auth_member!
 
+    skip_before_filter :set_gon
```
Comment (ysv) [inline app/controllers/admin/base_controller.rb]: Do we need this ?

File: app/views/layouts/admin.html.erb [hunk 5]
```diff
@@ -13,6 +13,7 @@
   <%= csrf_meta_tags %>
   <%= stylesheet_link_tag 'admin', media: 'all' %>
   <%= custom_stylesheet_link_tag_for :admin %>
+  <%= include_gon %>
```
Comment (ysv) [inline app/views/layouts/admin.html.erb]: What is the purpose of this code ?

**Review by ysv — CHANGES_REQUESTED** [review]
Add options to currencies.yml.erb usd because it brakes `rake db:seed` with nil options

File: app/controllers/admin/base_controller.rb (hunk 4 above)
Comment (dinesh-skyach) [reply app/controllers/admin/base_controller.rb]: @ysv Not required. I will remove this

File: app/views/layouts/admin.html.erb (hunk 5 above)
Comment (dinesh-skyach) [reply app/views/layouts/admin.html.erb]: @ysv Not required. I will remove this

**Comment by dinesh-skyach** [comment]
@ysv All changes done.

**Comment by ysv** [comment]
@dinesh-skyach it's still doesn't work properly https://vimeo.com/user92770823/review/306159684/591d8c64f0

**Review by ysv — APPROVED** [review]

**Comment by ysv** [comment]
@dinesh-skyach Open the same pr for 1-9 please

**Comment by dinesh-skyach** [comment]
@ysv ok

**Pull request merged**